src/components/ConnectWallet/ConnectWalletDialog.tsx

```diff
@@ -0,0 +1,229 @@
+  connection?: Web3Connection
+  onClick: () => void
+  onError?: (e: Error | undefined) => void
+}
+
+const wcQRUriAtom = atom<string | undefined>(undefined)
+
+function toQrCodeSvg(qrUri: string): Promise<string> {
+  return QRCode.toString(qrUri, {
+    // Leave a margin to increase contrast in dark mode.
+    margin: 1,
+    // Use 55*2=110 for the width to prevent distortion. The generated viewbox is "0 0 55 55".
+    width: 110,
+    type: 'svg',
+  })
+}
+
+function WalletConnectButton({ walletName, logoSrc, connection: wcTileConnection, onClick, onError }: ButtonProps) {
+  const [walletConnect] = wcTileConnection as [WalletConnect, Web3ReactHooks]
+  const defaultChainId = useAtomValue(defaultChainIdAtom)
+
+  const [qrUri, setQrUri] = useAtom(wcQRUriAtom)
+  const [qrCodeSvg, setQrCodeSvg] = useState<string>('')
+
+  useEffect(() => {
+    let stale = false
+    if (qrUri) {
+      toQrCodeSvg(qrUri).then((qrCodeSvg) => {
+        if (stale) return
+        setQrCodeSvg(qrCodeSvg)
+      })
+    } else {
+      walletConnect.activate(defaultChainId).catch((e) => {
+        if (stale) return
+        onError?.(e)
+      })
+    }
+    return () => {
+      stale = true
+    }
+  }, [qrUri, walletConnect, defaultChainId, onError])
+
+  useEffect(() => {
+    const disconnectListener = async (err: Error | null, _: any) => {
+      if (err) onError?.(err)
+      // Clear saved QR URI after disconnection
+      setQrUri(undefined)
+      walletConnect.deactivate()
+    }
+    walletConnect.provider?.connector.on('disconnect', disconnectListener)
+
+    // Need both URI event listeners
+    walletConnect.events.on(URI_AVAILABLE, async (uri: string) => {
+      if (uri) {
+        setQrUri(uri)
+      }
+    })
+
+    const uriListener = async (err: Error | null, payload: any) => {
+      if (err) onError?.(err)
+      const uri: string = payload.params[0]
+      if (uri) {
+        setQrUri(uri)
+      }
+    }
+    walletConnect.provider?.connector.on('display_uri', uriListener)
+
+    return () => {
+      walletConnect.events.off(URI_AVAILABLE)
+      ;(walletConnect.provider?.connector as unknown as EventEmitter | undefined)?.off('display_uri', uriListener)
+    }
+  })
+
+  return (
+    <StyledMainButton onClick={onClick}>
+      <StyledMainButtonRow>
+        <ButtonContents>
+          <img src={logoSrc} alt={walletName} width={32} />
+          <ThemedText.Subhead1>
+            <Trans>{walletName}</Trans>
+          </ThemedText.Subhead1>
+          <ThemedText.Caption color="secondary">
+            <Trans>Scan to connect your wallet. Works with most wallets.</Trans>
+          </ThemedText.Caption>
+        </ButtonContents>
+        <QRCodeWrapper dangerouslySetInnerHTML={{ __html: qrCodeSvg }} />
+      </StyledMainButtonRow>
+    </StyledMainButton>
+  )
+}
+
+function MetaMaskButton({ walletName, logoSrc, onClick }: ButtonProps) {
+  return (
+    <StyledSmallButton onClick={onClick}>
+      <ButtonContents>
```

Comment:
this buttoncontents stuff can be a component since it's also used in walletconnectbutton